Mathlib/Data/QuotLike.lean

```diff
@@ -0,0 +1,660 @@
+/-
+Copyright (c) 2024 Yuyang Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuyang Zhao
+-/
+import Mathlib.Logic.Relator
+import Mathlib.Logic.Unique
+import Qq.MetaM
+
+/-!
+# Typeclass for quotient types
+
+This typeclass is primarily for use by type synonyms of `Quot` and `Quotient`.
+
+Using `QuotLike` API for type synonyms of `Quot` and `Quotient` will avoid defeq abuse caused by
+directly using `Quot` and `Quotient` APIs.
+
+## Main definitions
+
+* `QuotLike Q α r`      : the type `Q` is canonically isomorphic to `Quot α r`.
+                          used for deriving `QuotLike` instances from the output type.
+* `QuotLike.HasQuot`    : used for deriving `QuotLike` instances from the input type.
+* `QuotLike.Hint`       : used for deriving `QuotLike` instances from the hint.
+* `QuotLike.HasQuotHint`: used for deriving `QuotLike` instances from the hint and the input type.
+* `mkQ`                 : the canonical quotient map inferred from the output type
+
+## Notations
+
+* `⟦a⟧`                 : the canonical quotient map inferred from the output type
+* `mkQ'` `⟦a⟧'`         : the canonical quotient map inferred from
+                            the input type via typeclass `QuotLike.HasQuot`
+* `mkQ_h` `⟦a⟧_h`       : the canonical quotient map inferred from
+                            the hint via typeclass `QuotLike.Hint`
+                            or the hint and the input type via typeclass `QuotLike.HasQuotHint`
+-/
+
+/--
+The definition of `QuotLike`. It is needed to avoid name clashes
+between `QuotLike.rec` `QuotLike.recOn` and the autogenerated recursors.
+-/
+class QuotLikeStruct (Q : Sort*) (α : outParam Sort*) (r : outParam (α → α → Prop)) where
+  /-- The canonical quotient map. Inferred from the output type. -/
+  mkQ : α → Q := by exact Quot.mk _
+  /-- The canonical map from quotient to `Quot r`. -/
+  toQuot : Q → Quot r := by exact (·)
+  toQuot_mkQ : ∀ a, toQuot (mkQ a) = Quot.mk r a := by exact fun _ ↦ rfl
+  /-- The analogue of `Quot.ind`: every element of `Q` is of the form `mkQ a`. -/
+  ind {motive : Q → Prop} : (∀ a : α, motive (mkQ a)) → ∀ q : Q, motive q := by exact Quot.ind
```

Comment:
I would require `Function.Surjective mkQ` instead (if you don't like the `Equiv` suggestion).